### PR TITLE
Add revocation support to path resolver

### PR DESCRIPTION
**Pull request by scrye**
This PR adds the necessary data structures to track IFIDs in paths. On SCMP revocation notifications, all paths containing the revoked IFID are flushed.

---
This change is [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion/1248)

### CODE REVIEW

**Comment by kormat** [comment]
The logic/dependencies/references in this are hard to follow. I had to start taking notes just to get a basic understanding of how this works. At the very least this needs some good documentation.

---

Reviewed 3 of 7 files at r1, 2 of 3 files at r2.
Review status: 5 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[go/lib/pathmgr/apppath.go, line 17 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp7S01oRHtkBJh6Avb:-Ktp7S01oRHtkBJh6Avc:bcqn9bv) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L17)):*
> ```Go
> package pathmgr
> 
> import "github.com/netsec-ethz/scion/go/lib/sciond"
> ```

Use `()` even if it's just a single import.

---

*[go/lib/pathmgr/apppath.go, line 59 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp93uDuEu89nSrtCN7:-Ktp93uDuEu89nSrtCN8:b-il3o6) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L59)):*
> ```Go
> // key returns a unique key that can be used for map indexing.
> func (ap *AppPath) key() rawKey {
> 	return rawKey(ap.Entry.Path.FwdPath)
> ```

That's.. a potentially huge key. Does it really need to be that large?

---

*[go/lib/pathmgr/defines.go, line 38 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp8r6OLkOdFdCVDa7u:-Ktp8r6OLkOdFdCVDa7v:bgl6c50) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/defines.go#L38)):*
> ```Go
> 
> func (r rawKey) String() string {
> 	return (common.RawBytes)(r).String()
> ```

`common.RawBytes(r).String()`

---

*[go/lib/pathmgr/pathmgr.go, line 128 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp9xATQBrVeHlHlxvN:-Ktp9xATQBrVeHlHlxvO:bu300d2) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L128)):*
> ```Go
> 	if ok {
> 		pathSet := pathSetI.(AppPathSet)
> 		return pathSet
> ```

This can be simplified to `return pathSetI.(AppPathSet)`

---

*[go/lib/pathmgr/pathmgr.go, line 220 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpAJaT-NXIChFWlEDf:-KtpAJaT-NXIChFWlEDg:b5yx7tl) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L220)):*
> ```Go
> 	case sciond.RevUnknown:
> 		fallthrough
> 	case sciond.RevValid:
> ```

`case sciond.RevUnknown, sciond.RevValid:`

---

*[go/lib/pathmgr/pathmgr.go, line 222 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpGHAiOSLlnylLjeCf:-KtpGHAiOSLlnylLjeCg:biyynrk) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L222)):*
> ```Go
> 	case sciond.RevValid:
> 		uifid := UIFIDFromValues(parsedRev.IA(), common.IFIDType(parsedRev.IfID))
> 		r.revTable.revoke(uifid)
> ```

Entries in `pathMap` and `SyncPaths` aren't revoked.

---

*[go/lib/pathmgr/pathmgr.go, line 273 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpAcVZl-t1CVgVvH6S:-KtpAcVZl-t1CVgVvH6T:b-t221ls) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L273)):*
> ```Go
> 
> 	aps := NewAppPathSet(reply)
> 	return aps
> ```

`return NewAppPathSet(reply)`

---

*[go/lib/pathmgr/uifid.go, line 40 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp6tOMnmnN0LtgyKD0:-Ktp6tONmbkyK6qsMSQa:b-87p915) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L40)):*
> ```Go
> // key returns a unique key that can be used as a map index
> func (u *UIFID) key() string {
> 	return fmt.Sprintf("%v.%d", u.isdas, u.ifid)
> ```

`%s#%d`.

---

*[go/lib/pathmgr/uifid.go, line 51 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpFhYQK9khi_6S2PQO:-KtpFhYRPJH5b6DG-b65:by4wnr5) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L51)):*
> ```Go
> // and calling Revoke on each AppPath.
> type revTable struct {
> 	m map[string]AppPathSet
> ```

As the key is never referenced (afaict), this could just be a []AppPathSeg.

---

*[go/lib/pathmgr/uifid.go, line 68 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpCkq1ecgh_Fy7tU48:-KtpCkq1ecgh_Fy7tU49:b-fgmmyw) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L68)):*
> ```Go
> // structures. This allows the RevTable to track multiple copies of the same
> // path if needed.
> func (u *revTable) updatePathSet(aps AppPathSet, disc int) {
> ```

`u` is a non-obvious name for the receiver.

---

*[go/lib/pathmgr/uifid.go, line 78 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpBSuznESLF6UFhaGu:-KtpBSuznESLF6UFhaGv:b9025m6) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L78)):*
> ```Go
> 		uifid := UIFIDFromValues(iface.ISD_AS(), common.IFIDType(iface.IfID))
> 		aps := u.m[uifid.key()]
> 		if aps == nil {
> ```

A more conventional way to do this is `aps, ok := ...; if !ok {`. That will also differentiate between "key not in map", and "value in map is `nil`".

---

*[go/lib/pathmgr/uifid.go, line 92 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpG2zJfPm8LqYprLHr:-KtpG2zJfPm8LqYprLHs:b-jft2nl) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L92)):*
> ```Go
> 	aps := u.m[uifid.key()]
> 	for _, ap := range aps {
> 		ap.revoke()
> ```

Hurm. This will only remove a tiny proportion of references to a given AppPath. All of the other revTable references for all the other interfaces it goes through aren't cleaned up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-:-KtpGWX63BSPAlPhGHmV:bx1iond)*

**Comment by scrye** [comment]
Can you point me in the direction of what was difficult to follow? I am more than willing to improve documentation, but I'd like to better understand what to focus on.

---

Review status: 1 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[go/lib/pathmgr/apppath.go, line 17 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp7S01oRHtkBJh6Avb:-KtpJbcxgR9lgdTxc7V1:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L17)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Use `()` even if it's just a single import.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/apppath.go, line 59 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp93uDuEu89nSrtCN7:-KtpJwvTfCkXCLPVHzvU:bw5zfmp) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L59)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

That's.. a potentially huge key. Does it really need to be that large?
</blockquote></details>

Internally, Go maps are just hashtables. The string is hashed behind the scenes to a small one. Compressing it ourselves to something smaller wouldn't give any benefits but would increase complexity.

Source: https://golang.org/src/runtime/hashmap.go

---

*[go/lib/pathmgr/defines.go, line 38 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp8r6OLkOdFdCVDa7u:-KtpOHyw3kVVjIpcQAEH:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/defines.go#L38)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`common.RawBytes(r).String()`
</blockquote></details>

Done.

---

*[go/lib/pathmgr/pathmgr.go, line 128 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp9xATQBrVeHlHlxvN:-KtpORg9-LlOXmPuzSBJ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L128)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This can be simplified to `return pathSetI.(AppPathSet)`
</blockquote></details>

Done.

---

*[go/lib/pathmgr/pathmgr.go, line 220 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpAJaT-NXIChFWlEDf:-KtpOSnJcSafWFf7C29-:bhqdubz) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L220)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`case sciond.RevUnknown, sciond.RevValid:`
</blockquote></details>

Huh, I never knew one can do that. Good to know :D

---

*[go/lib/pathmgr/pathmgr.go, line 222 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpGHAiOSLlnylLjeCf:-KtpOlL9X7eubzBSA5MV:bjk2fiw) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L222)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Entries in `pathMap` and `SyncPaths` aren't revoked.
</blockquote></details>

Conclusion to offline discussion: They are revoked, but they aren't garbage collected.

Added new code to cleanup everything, also updated tests to check that the references are deleted.

---

*[go/lib/pathmgr/pathmgr.go, line 273 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpAcVZl-t1CVgVvH6S:-KtplFfxB-bZzRfcwuLV:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/pathmgr.go#L273)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`return NewAppPathSet(reply)`
</blockquote></details>

Done.

---

*[go/lib/pathmgr/uifid.go, line 40 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp6tOMnmnN0LtgyKD0:-KtplNQGQl2X-bzuNJ54:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L40)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`%s#%d`.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/uifid.go, line 51 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpFhYQK9khi_6S2PQO:-KtplRq7iIQ1H8CbcBkJ:b-d25rt2) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L51)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

As the key is never referenced (afaict), this could just be a []AppPathSeg.
</blockquote></details>

The key is used during `revoke` to jump directly to the relevant `AppPathSet` that contains all the paths that contain that specific UIFID.

---

*[go/lib/pathmgr/uifid.go, line 68 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpCkq1ecgh_Fy7tU48:-Ktplg1N3mXjO08IBdKq:b-i2hrn8) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L68)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`u` is a non-obvious name for the receiver.
</blockquote></details>

Refactoring fail. Changed it now.

---

*[go/lib/pathmgr/uifid.go, line 78 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpBSuznESLF6UFhaGu:-KtpnNjFBKfe1koY-Mer:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L78)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

A more conventional way to do this is `aps, ok := ...; if !ok {`. That will also differentiate between "key not in map", and "value in map is `nil`".
</blockquote></details>

Done.

---

*[go/lib/pathmgr/uifid.go, line 92 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtpG2zJfPm8LqYprLHr:-KtpnSZkmpWxcAI3XDrl:b-1hxwtx) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/uifid.go#L92)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Hurm. This will only remove a tiny proportion of references to a given AppPath. All of the other revTable references for all the other interfaces it goes through aren't cleaned up.
</blockquote></details>

Added new code to cleanup everything, also updated tests to check that the references are deleted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-:-KtpncLHphIoF64QOQD9:btos9iv)*

**Comment by kormat** [comment]
How these all fit together:
- pathMap
- SyncPath
- AppPathSeg
- AppPath
- revTable
How they are used during lookup, and during revocation.

---

Reviewed 1 of 4 files at r2, 5 of 7 files at r3.
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[go/lib/pathmgr/apppath.go, line 59 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp93uDuEu89nSrtCN7:-Ktq4fAHwPV0Krf3FxeH:bu77wlo) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L59)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Internally, Go maps are just hashtables. The string is hashed behind the scenes to a small one. Compressing it ourselves to something smaller wouldn't give any benefits but would increase complexity.

Source: https://golang.org/src/runtime/hashmap.go
</blockquote></details>

We don't need to compress it, as such. Just give each `AppPath` an `Id` field, which could be a sequential int, or even the request id used for sciond. That could be used as the key in various places.

---

*[go/lib/pathmgr/apppath_test.go, line 27 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqJrLBNtNzlXQ5dQEd:-KtqJrLBNtNzlXQ5dQEe:bt3zsjr) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L27)):*
> ```Go
> var (
> 	paths = map[string]*sciond.PathReply{
> 		"1-19.2-25": {
> ```

There are 8 paths here, but it looks like only 3 are used. 

Ah, some are used from the other _test file. In that case, please add a comment here mentioning this.

---

*[go/lib/pathmgr/apppath_test.go, line 42 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqLL44b9aLKtQtdvW3:-KtqLL44b9aLKtQtdvW4:b-qltcud) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L42)):*
> ```Go
> 						Mtu: 0x500,
> 						Interfaces: []sciond.PathInterface{
> 							{RawIsdas: 0x100013, IfID: 0x3c}, {RawIsdas: 0x100010, IfID: 0x26},
> ```

The ifids are referred to using decimal elsewhere, which makes this hard to correlate. I suggest using decimal here, too, as that's also what the strings use.

The same goes for the raw IA numbers. They could be replaced with e.g. `addr.IAFromString("1-10").Uint32()`, or a small helper function which returns the result.

---

*[go/lib/pathmgr/apppath_test.go, line 55 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqJ1yy8xky2wMIe3He:-KtqJ1yy8xky2wMIe3Hf:bmgj598) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L55)):*
> ```Go
> 						}{
> 							Ipv4: []uint8{0x7f, 0x0, 0x0, 0xb1},
> 							Ipv6: []uint8(nil)}}}}},
> ```

Stacking closing }'s like this makes it very hard to follow scopes.

---

*[go/lib/pathmgr/pathmgr.go, line 215 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqHxjaNfaTtr7tbShG:-KtqHxjaNfaTtr7tbShH:b2foka8) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/pathmgr.go#L215)):*
> ```Go
> 		uifid := UIFIDFromValues(parsedRev.IA(), common.IFIDType(parsedRev.IfID))
> 		r.revTableCache.revoke(uifid)
> 		r.revTableReg.revoke(uifid)
> ```

If a revocation removes all paths for a registered destination, it should immediately trigger a new lookup to be done.

---

*[go/lib/pathmgr/revtable_test.go, line 31 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqKqjYtbVr0jgnRntg:-KtqKqjYtbVr0jgnRnth:b-w76mhq) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/revtable_test.go#L31)):*
> ```Go
> 		SoMsg("revocation table", rt, ShouldNotEqual, nil)
> 
> 		Convey("Add path set for 1-19.2-25", func() {
> ```

While we don't have a _standard_, this is normally represented as "1-19 -> 2-25".

---

*[go/lib/pathmgr/revtable_test.go, line 36 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqKyF4pGm3PuUZSbWa:-KtqKyF4pGm3PuUZSbWb:b-5sanx3) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/revtable_test.go#L36)):*
> ```Go
> 			SoMsg("revocation table size", len(rt.m), ShouldEqual, 12)
> 
> 			Convey("Revoke non-existing UIFID 1-18.42", func() {
> ```

`1-18#42` is the proper notation for this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-:-Ktq58jUv25BFKNDiTUL:b-humd5o)*

**Comment by scrye** [comment]
Review status: 6 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[go/lib/pathmgr/apppath.go, line 59 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Ktp93uDuEu89nSrtCN7:-KtqOzDMp9CcLhRHhVQ1:b-783vli) ([raw file](https://github.com/netsec-ethz/scion/blob/0ff50351b4e227506cf53c8f7475424713ec7dcc/go/lib/pathmgr/apppath.go#L59)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

We don't need to compress it, as such. Just give each `AppPath` an `Id` field, which could be a sequential int, or even the request id used for sciond. That could be used as the key in various places.
</blockquote></details>

Changed key function for AppPaths to use the `sha256` of IAs and IFIDs.

---

*[go/lib/pathmgr/apppath_test.go, line 27 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqJrLBNtNzlXQ5dQEd:-Ktqfl7zn6zeYEzRblGI:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L27)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

There are 8 paths here, but it looks like only 3 are used. 

Ah, some are used from the other _test file. In that case, please add a comment here mentioning this.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/apppath_test.go, line 42 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqLL44b9aLKtQtdvW3:-KtqfPNVKMkdDZ_qq4O3:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L42)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

The ifids are referred to using decimal elsewhere, which makes this hard to correlate. I suggest using decimal here, too, as that's also what the strings use.

The same goes for the raw IA numbers. They could be replaced with e.g. `addr.IAFromString("1-10").Uint32()`, or a small helper function which returns the result.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/apppath_test.go, line 55 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqJ1yy8xky2wMIe3He:-KtqbigN7P04NmPQmujr:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/apppath_test.go#L55)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Stacking closing }'s like this makes it very hard to follow scopes.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/pathmgr.go, line 215 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqHxjaNfaTtr7tbShG:-KtqpTpb7QIhXBCsy1rI:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/pathmgr.go#L215)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

If a revocation removes all paths for a registered destination, it should immediately trigger a new lookup to be done.
</blockquote></details>

Done.

---

*[go/lib/pathmgr/revtable_test.go, line 31 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqKqjYtbVr0jgnRntg:-Ktqpb7sUoHZmejYda3c:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/revtable_test.go#L31)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

While we don't have a _standard_, this is normally represented as "1-19 -> 2-25".
</blockquote></details>

Done.

---

*[go/lib/pathmgr/revtable_test.go, line 36 at r3](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-KtqKyF4pGm3PuUZSbWa:-Ktqpk2oQ47v2WOTqpmJ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/9a03e881a842b78c0fccfe472cb0f78287df6222/go/lib/pathmgr/revtable_test.go#L36)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`1-18#42` is the proper notation for this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248)*

**Comment by kormat** [comment]
Reviewed 2 of 7 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[go/lib/pathmgr/revtable.go, line 104 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Kus3DHmUhGsXFhF2wud:-Kus3DHmUhGsXFhF2wue:bt7ytvj) ([raw file](https://github.com/netsec-ethz/scion/blob/8845dc0ae6e8ecc03970bed8f296b9832cbd14d8/go/lib/pathmgr/revtable.go#L104)):*
> ```Go
> 			src := getSrcIA(ap)
> 			dst := getDstIA(ap)
> 			pairs = append(pairs, &IAPair{src: src, dst: dst})
> ```

`pairs = append(pairs, &IAPair{src: getSrcIA(ap), dst: getDstIA(ap)})`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248)*

**Comment by scrye** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/lib/pathmgr/revtable.go, line 104 at r4](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-Kus3DHmUhGsXFhF2wud:-Kv2GdTW0zN4zLlczGS4:b-896fix) ([raw file](https://github.com/netsec-ethz/scion/blob/8845dc0ae6e8ecc03970bed8f296b9832cbd14d8/go/lib/pathmgr/revtable.go#L104)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`pairs = append(pairs, &IAPair{src: getSrcIA(ap), dst: getDstIA(ap)})`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248)*

**Comment by kormat** [comment]
:lgtm:

---

Reviewed 1 of 7 files at r3, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion/1248#-:-Kv2Qs5DtSIHntEZXAWM:bnfp4nl)*
